include/rmm/mr/device/detail/free_list.hpp

```diff
@@ -33,9 +33,9 @@ namespace detail {
  *        of memory allocated from the heap (or upstream allocator).
  */
 struct block {
-  char* ptr;     ///< Raw memory pointer
-  size_t size;   ///< Size in bytes
-  bool is_head;  ///< Indicates whether ptr was allocated from the heap
+  block() = default;
+  explicit block(char* ptr) : ptr{ptr}, size(0), is_head(true) {}
```

Comment:
When can you have a pointer other than `nullptr` when the `size` is zero?

Reply:
When you need to find an allocated block to free it when you only know its pointer, as done here: 
https://github.com/harrism/rmm/blob/79f077dbbe852684269bf98dfe6efea6df85d50a/include/rmm/mr/device/pool_memory_resource.hpp#L233

Reply:
In that example, you know what the size is. I know that it's not useful for the `find` operation, but it doesn't feel right to allow constructing a `block` from a non-null ptr with a zero size. Could lead to bugs later down the road where you construct a malformed block without knowing it.

Reply:
I discovered that C++14 provides an elegant solution to this problem: `is_transparent`. So I defined a generic `block_comparator` struct so now the set of blocks can be searched by pointer rather than by block. This eliminates the need for the single-argument constructor. All is well.

Reply:
Wow, it's really cool when you discover a solution that was designed to solve your exact problem.